### PR TITLE
Config logic

### PR DESCRIPTION
**Pull request by JRMeyer**
This is a draft PR because I know the logic isn't pretty at the moment, especially with regard to `_SttConfig` and `Config`

The reason for this PR is to break the assumption of CLI args in `train.py` and deeper, in `coqui_stt_training.util.config.initialize_globals`

I'd like to make training with STT easier programmatically, within python, not just from the CLI

### CODE REVIEW

File: lm_optimizer.py [hunk 1]
```diff
@@ -52,7 +52,8 @@ def objective(trial):
 
 
 def main(_):
-    initialize_globals()
+    Config = _SttConfig()
+    initialize_config_globals(Config)
```
Comment (erogol) [inline lm_optimizer.py:56]: It looks like you can implement the `initialize_config_globals` in the `def _post_init_()` function of `_SttConfig` if you want to make things even shorter.

File: training/coqui_stt_training/util/config.py [hunk 2]
```diff
@@ -546,9 +546,10 @@ def check_values(self):
         check_argument("one_shot_infer", c, is_path=True)
 
 
-def initialize_globals():
-    c = _SttConfig()
-    c.parse_args(arg_prefix="")
+def initialize_config_globals(c):
```
Comment (erogol) [inline training/coqui_stt_training/util/config.py:549]: It looks like you can implement the initialize_config_globals in the def _post_init_() function of _SttConfig if you want to make things even shorter.

**Review by JRMeyer — COMMENTED** [review]
@erogol what about the case where I have args I want to declare in a script (e.g. a jupyter notebook)?

Currently I have something like this:

```
Config = _SttConfig()

Config.alphabet = Alphabet('/home/STT/data/alphabet.txt')
Config.train_files=['/home/STT/data/ldc93s1.csv']
Config.dev_files=['/home/STT/data/ldc93s1.csv']
Config.test_files=['/home/STT/data/ldc93s1.csv']
Config.n_hidden=100
Config.epochs=200

initialize_config_globals(Config)
```

**Comment by erogol** [comment]
@JRMeyer it does not stop you from doing that. 
https://docs.python.org/3/library/dataclasses.html#post-init-processing

So you init the variable, then update them with a certain logic in the `_post_init_` then you change the ones you want in your script. 

So the end result would be 

```python
Config = _SttConfig()

Config.alphabet = Alphabet('/home/STT/data/alphabet.txt')
Config.train_files=['/home/STT/data/ldc93s1.csv']
Config.dev_files=['/home/STT/data/ldc93s1.csv']
Config.test_files=['/home/STT/data/ldc93s1.csv']
Config.n_hidden=100
Config.epochs=200
```

Effectively, `initialize_config_globals` is embedded into `_post_init_`

The only limitation I see is that if the logic in `initialize_config_globals` depends on the values changed in the script then you need to call `_post_init_`once again.

**Review by reuben — APPROVED** [review]
r+ with tests passing

File: bin/run-ci-ldc93s1_new_bytes_tflite.sh [hunk 3]
```diff
@@ -16,7 +16,7 @@ fi;
 # and when trying to run on multiple devices (like GPUs), this will break
 export CUDA_VISIBLE_DEVICES=0
 
-python -u train.py --show_progressbar false \
+python -u train.py --alphabet_config_path "data/alphabet.txt" --show_progressbar false \
```
[the remark this answers is not in the document]
Comment (JRMeyer) [reply bin/run-ci-ldc93s1_new_bytes_tflite.sh]: fixed in 0389560a927a5c692e3747e7f4beac27e7069bb6

File: bin/run-ci-ldc93s1_new_bytes.sh [hunk 4]
```diff
@@ -17,7 +17,7 @@ fi;
 # and when trying to run on multiple devices (like GPUs), this will break
 export CUDA_VISIBLE_DEVICES=0
 
-python -u train.py --show_progressbar false --early_stop false \
+python -u train.py --alphabet_config_path "data/alphabet.txt" --show_progressbar false --early_stop false \
```
[the remark this answers is not in the document]
Comment (JRMeyer) [reply bin/run-ci-ldc93s1_new_bytes.sh]: fixed in 0389560a927a5c692e3747e7f4beac27e7069bb6

File: bin/run-ci-ldc93s1_checkpoint_bytes.sh [hunk 5]
```diff
@@ -14,7 +14,7 @@ fi;
 # and when trying to run on multiple devices (like GPUs), this will break
 export CUDA_VISIBLE_DEVICES=0
 
-python -u train.py --show_progressbar false --early_stop false \
+python -u train.py --alphabet_config_path "data/alphabet.txt" --show_progressbar false --early_stop false \
```
[the remark this answers is not in the document]
Comment (JRMeyer) [reply bin/run-ci-ldc93s1_checkpoint_bytes.sh]: fixed in 0389560a927a5c692e3747e7f4beac27e7069bb6

**Pull request merged**